test/helm_task_test.go

```diff
@@ -0,0 +1,310 @@
+		c := pr.Status.GetCondition(duckv1alpha1.ConditionSucceeded)
+		if c != nil {
+			if c.Status == corev1.ConditionTrue {
+				return true, nil
+			} else if c.Status == corev1.ConditionFalse {
+				return true, fmt.Errorf("pipeline run %s failed!", helmDeployPipelineRunName)
+			}
+		}
+		return false, nil
+	}, "PipelineRunCompleted"); err != nil {
+		t.Errorf("Error waiting for PipelineRun %s to finish: %s", helmDeployPipelineRunName, err)
+	}
+
+	k8sService, err := c.KubeClient.Kube.CoreV1().Services(namespace).Get(helmDeployServiceName, metav1.GetOptions{})
+	if err != nil {
+		t.Errorf("Error getting service at %s %s", helmDeployServiceName, err)
+	}
+	var serviceIp string
+	ingress := k8sService.Status.LoadBalancer.Ingress
+	if len(ingress) > 0 {
+		serviceIp = ingress[0].IP
+	}
+
+	resp, err := http.Get(fmt.Sprintf("http://%s:8080", serviceIp))
+	if err != nil {
+		t.Errorf("Error reaching service at http://%s:8080 %s", serviceIp, err)
+	}
+	if resp != nil && resp.StatusCode != http.StatusOK {
+		t.Errorf("Error from service at http://%s:8080 %s", serviceIp, err)
+	}
+}
+
+func getGoHelloworldGitResource(namespace string) *v1alpha1.PipelineResource {
+	return &v1alpha1.PipelineResource{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      sourceResourceName,
+			Namespace: namespace,
+		},
+		Spec: v1alpha1.PipelineResourceSpec{
+			Type: v1alpha1.PipelineResourceTypeGit,
+			Params: []v1alpha1.Param{
+				v1alpha1.Param{
+					Name:  "Url",
+					Value: "https://github.com/pivotal-nader-ziada/gohelloworld",
+				},
+			},
+		},
+	}
+}
+
+func getCreateImageTask(namespace string, t *testing.T) *v1alpha1.Task {
+	// according to knative/test-infra readme (https://github.com/knative/test-infra/blob/13055d769cc5e1756e605fcb3bcc1c25376699f1/scripts/README.md)
+	// the KO_DOCKER_REPO will be set with according to the porject where the cluster is created
+	// it is used here to dunamically get the docker registery to push the image to
+	dockerRepo := os.Getenv("KO_DOCKER_REPO")
+	if dockerRepo == "" {
+		t.Fatalf("KO_DOCKER_REPO env variable is required")
+	}
+
+	imageName = fmt.Sprintf("%s/%s", dockerRepo, sourceImageName)
+
+	return &v1alpha1.Task{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: namespace,
+			Name:      createImageTaskName,
+		},
+		Spec: v1alpha1.TaskSpec{
+			Inputs: &v1alpha1.Inputs{
+				Resources: []v1alpha1.TaskResource{
+					v1alpha1.TaskResource{
+						Name: sourceResourceName,
+						Type: v1alpha1.PipelineResourceTypeGit,
+					},
+				},
+			},
+			BuildSpec: &buildv1alpha1.BuildSpec{
+				Steps: []corev1.Container{{
+					Name:  "kaniko",
+					Image: "gcr.io/kaniko-project/executor",
+					Args: []string{"--dockerfile=/workspace/Dockerfile",
+						fmt.Sprintf("--destination=%s", imageName),
+					},
+				}},
+			},
+		},
+	}
+}
+
+func getHelmDeployTask(namespace string) *v1alpha1.Task {
+	return &v1alpha1.Task{
```

Comment:
This is something, we haven't figured out yet. You can address it now or later.
Shd we host all the Tasks Defn somewhere for others to use? 
And then the tests should read the task def from that file. 
The reason why i think we shd do so is
- Users can see all the tasks def at one place 
-  In future, if we decide to make them public-ally available, there are no two copies and tests always test the public ones.

Reply:
I thought about this and here are my reason for keeping these tasks in this test:
 * the generic file with all tasks will grow too big and be hard to read
  * these tasks should be specific to the test using them and having them in a common file can make changing tests harder
  * I expect the public task definitions to be in yaml and in a different repo all together

Reply:
you make a valid point, its open for discussion for sure

Reply:
We can keep iterating on this but what I found when working on knative/serving conformance tests was that using yaml directly was hard in tests b/c sometimes we'd need a field to be dynamic (e.g. namespaces)

So what I'd like to do in the long run is have both integration tests like these that use resources defined in code which cover very specific cases we want to test, and examples in `examples` which are yaml and tested (see https://github.com/knative/build-pipeline/issues/108).

I think you're right @tejal29 , we'll probably remove some of these integration test cases in the long run (maybe even just one, with a complex pipeline).

Reply:
sounds good to me. We shd figure out a plan on where and how to make tasks available. 
Something like [krew](https://github.com/GoogleContainerTools/krew)
And then a plan to promote well tested tasks from here to this repo/location.
Also make sure, we don't make any backward incompatible changes to the controller and break existing tasks.